Ensure that a `@_unsafeInheritExecutor` function does not depend on `#isolation`

DougGregor opened the pull request:
An `@_unsafeInheritExecutor` function is unsafe because it doesn't really "inherit" the executor, it just avoids immediately hopping off the executor. That means that using `#isolation` within such a function is fundamentally broken. Ban the use of `#isolation` within such a function, providing a Fix-It that removes the `@_unsafeInheritExecutor` attribute and adds a defaulted parameter

    isolation: (any Actor)? = #isolation

instead. That's the real, safe pattern that we want going forward. We take the extra step of diagnosing `@_unsafeInheritExecutor` as an error in Swift 6, downgraded to a warning in Swift 5. There's no place for this misfeature going forward, since we have a proper solution in the language.

Unfortunately, the async `for..in` loop injects an `#isolation` when used with the `next(isolation:)`, so avoid using `next(isolation:)` when inside a `@_unsafeInheritExecutor` function, because that would immediately break.

Even more unfortunately, the move from `@_unsafeInheritExecutor` to `#isolation` for the with*Continuation functions in the concurrency library breaks code that is using `@_unsafeInheritExecutor` and calling these APIs. This currently causes silent breakage (which manifest as runtime crashes), and with this change detected by the compiler as an error. However, we don't want to outright break such code, so we have a workaround. It's ugly, and you might want to stop reading now.

Introduce _unsafeInheritExecutor_-prefixed versions of the `with*Continuation` and `withTaskCancellationHandler` APIs into
the _Concurrency library that use `@_unsafeInheritExecutor`. Then, teach the type checker to swap in these `_unsafeInheritExecutor_`-prefixed versions in lieu of the originals when they are called from an `@_unsafeInheritExecutor` function. This allows existing code using `@_unsafeInheritExecutor` with these APIs to continue working as it has before, albeit with a warning that `@_unsafeInheritExecutor` has been removed.

Fixes rdar://131151376.

DougGregor committed:
Diagnose attempts to use #isolation within an `@_unsafeInheritExecutor` function

An `@_unsafeInheritExecutor` function is unsafe because it doesn't
really "inherit" the executor, it just avoids immediately hopping off
the executor. That means that using `#isolation` within such a
function is fundamentally broken. Ban the use of `#isolation` within
such a function, providing a Fix-It that removes the
`@_unsafeInheritExecutor` attribute and adds a defaulted parameter

    isolation: (any Actor)? = #isolation

instead. That's the real, safe pattern that want going forward.

We did say it was unsafe, after all. Part of rdar://131151376.

DougGregor committed:
Avoid using `AsyncIteratorProtocol.next(isolation:)` with `@_unsafeInheritExecutor`

`#isolation` in unsafe within a function that uses
`@_unsafeInheritExecutor`, so avoid the implicit use of the new
`AsyncIteratorProtocol.next(isolation:)` in such functions.

DougGregor committed:
Deprecate `@_unsafeInheritExecutor` and remove it from Swift 6

DougGregor committed:
Use `@_unsafeInheritExecutor` forms of with*Continuation from `@_unsafeInheritExecutor` functions

The move from `@_unsafeInheritExecutor` to `#isolation` for the
with*Continuation breaks code that is using `@_unsafeInheritExecutor` and
calling these APIs. This originally caused silent breakage (which manifest
as runtime crashes), and is now detected by the compiler as an error.

However, despite `@_unsafeInheritExecutor` being an unsafe,
not-intended-to-be-user-facing feature, it is indeed being used, along
with these APIs. Introduce _unsafeInheritExecutor_-prefixed versions of
the `with*Continuation` and `withTaskCancellationHandler` APIs into
the _Concurrency library that use `@_unsafeInheritExecutor`. Then,
teach the type checker to swap in these
_unsafeInheritExecutor_-prefixed versions in lieu of the originals
when they are called from an `@_unsafeInheritExecutor` function. This
allows existing code using `@_unsafeInheritExecutor` with these APIs
to continue working as it has before, albeit with a warning that
`@_unsafeInheritExecutor` has been removed.

Fixes rdar://131151376.

DougGregor commented:
@swift-ci please smoke test

ktoso commented on lib/Sema/TypeCheckAttr.cpp:
```diff
@@ -7279,6 +7279,13 @@ void AttributeChecker::visitUnsafeInheritExecutorAttr(
+    bool inConcurrencyModule = D->getDeclContext()->getParentModule()->getName()
+        .str().equals("_Concurrency");
+    auto diag = fn->diagnose(diag::unsafe_inherits_executor_deprecated);
+    diag.warnUntilSwiftVersion(6);
```
I'm a bit worried about making it a hard error since I've not been so confident in all the situations with isolation inference and #isolation actually working just yet... but yes, long term this should be an error. 

Perhaps we just have to take the risk.

DougGregor replied:
Right. The combination of unsafe / underscored / does not compose with its replacement is what pushed me to outright ban it in Swift 6, but we could back off from this.

ktoso replied:
It’s probably fine, worst that can happen in oss is delaying 6 mode adoption in case #isolation still has some bugs. That’s okey I think, worth pushing folks off the underscored thing because it’s so broken

ktoso commented on stdlib/public/Concurrency/CheckedContinuation.swift:
```diff
@@ -300,11 +300,10 @@ public func withCheckedContinuation<T>(
 @_unavailableInEmbedded
 @_silgen_name("$ss23withCheckedContinuation8function_xSS_yScCyxs5NeverOGXEtYalF")
-internal func __abi_withCheckedContinuation<T>(
+public func _unsafeInheritExecutor_withCheckedContinuation<T>(
```
We've been pretty consistent with keeping all those methods "kept around for abi reasons" with the `__abi_` prefix, is there a real reason to change away from that here?

ktoso replied:
It could be some __abi_unsafe... if you prefer, but it's easy to skim and notice why the hell such methods exist by looking at prefix, so I'd keep them tbh

DougGregor replied:
I can add more comments here. The `_unsafeInheritExecutor_` prefix is load-bearing in the awful, awful hack we're using to transitionally from `@_unsafeInheritExecutor`.

ktoso replied:
The more comments help enough here, Lgtm thanks :)

ktoso commented on stdlib/public/Concurrency/PartialAsyncTask.swift:
```diff
@@ -730,6 +730,32 @@ public func withUnsafeThrowingContinuation<T>(
+@_unsafeInheritExecutor
+public func _unsafeInheritExecutor_withUnsafeContinuation<T>(
+  _ fn: (UnsafeContinuation<T, Never>) -> Void
+) async -> T {
```
This should be `sending T`, shouldn't it? At the very least, its now inconsistent between throwing and not throwing which seems off. 

And note that both our "good" APIs are `sending`


```
@_alwaysEmitIntoClient
public func withUnsafeContinuation<T>(
  isolation: isolated (any Actor)? = #isolation,
  _ fn: (UnsafeContinuation<T, Never>) -> Void
) async -> sending T {

@_alwaysEmitIntoClient
public func withUnsafeThrowingContinuation<T>(
  isolation: isolated (any Actor)? = #isolation,
  _ fn: (UnsafeContinuation<T, Error>) -> Void
) async throws -> sending T {
```

DougGregor replied:
Good catch, thank you!

ktoso commented on test/Concurrency/unsafe_inherit_executor.swift:
```diff
@@ -32,10 +34,78 @@ class NonSendableObject {
+
+  await inheritsIsolationProperly()
+  // expected-error@-1{{#isolation (introduced by a default argument) cannot be used within an '@_unsafeInheritExecutor' function}}{{50:1-24=}}{{51:26-26=, isolation: isolated (any Actor)? = #isolation}}
+t 6)
```
looks like typo that'll fail to build?

ktoso commented on stdlib/public/Concurrency/PartialAsyncTask.swift:
```diff
@@ -730,6 +730,32 @@ public func withUnsafeThrowingContinuation<T>(
 }
 
+// HACK: a version of withUnsafeContinuation that uses the unsafe
+// @_unsafeInheritExecutor.
```
Would you mind adding a "why" here, it can be difficult remembering why all those `HACK:` exist sometimes 😉 

 ```suggestion
// @_unsafeInheritExecutor. The type checker will automatically replace calls to 
// #isolation-based methods to their `__abi__unsafeInheritExecutor_` version
// if called from an @_unsafeInheritExecutor context.
```

DougGregor replied:
I'm switching to this boilerplate:

```
// Note: hack to stage out @_unsafeInheritExecutor forms of various functions
// in favor of #isolation. The _unsafeInheritExecutor_ prefix is meaningful
// to the type checker.
//
// This function also doubles as an ABI-compatibility shim predating the
// introduction of #isolation.
```

ktoso replied:
Sounds good to me, thank you!

ktoso reviewed (approved):
Looks good though some minor comments inline.

DougGregor committed:
Address code review feedback and fix rebase typo in a test

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please smoke test Windows

hborla reviewed (approved):
The pre-check expression changes LGTM. I'm also in favor of banning `@_unsafeInheritExecutor` in Swift 6 mode. The only reason I can think of to continue to use it in Swift 6 mode to stage in the isolated parameter to an existing API in a resilient library (like we've done in the Concurrency library), which we could permit more narrowly where necessary.

DougGregor committed:
Add _unsafeInheritExecutor_ version of TaskLocal.withValue

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

ktoso reviewed (approved):
LGTM on the _unsafeInheritExecutor_withValue as well

DougGregor committed:
Extend type checker hack for `@_unsafeInheritExecutor` functions to methods

With the re-introduction of `@_unsafeInheritExecutor` for `TaskLocal.withValue`,
we need to extend the type checker trick with `_unsafeInheritExecutor_`-prefixed
functions to work with methods. Do so to make `TaskLocal.withValue` actually
work this way.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please smoke test Linux
